src/protocols/bdx/tests/TestBdxTransferSession.cpp

```diff
@@ -0,0 +1,335 @@
+{
+    // Verify that receiver emits BlockQuery message
+    CHIP_ERROR err = receiver.PrepareBlockQuery();
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+    receiver.PollOutput(outEvent);
+    NL_TEST_ASSERT(inSuite, outEvent.EventType == TransferSession::kOutput_MsgToSend);
+    System::PacketBufferHandle queryMsgToSend = outEvent.MsgData.Retain();
+    NL_TEST_ASSERT(inSuite, !queryMsgToSend.IsNull());
+
+    // Pass BlockQuery to sender
+    err = sender.HandleMessageReceived(queryMsgToSend.Retain());
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+
+    // Verify sender emits QueryReceived event
+    sender.PollOutput(outEvent);
+    NL_TEST_ASSERT(inSuite, outEvent.EventType == TransferSession::kOutput_QueryReceived);
+}
+
+void SendAndVerifyArbitraryBlock(nlTestSuite * inSuite, void * inContext, TransferSession & sender, TransferSession & receiver,
+                                 TransferSession::OutputEvent & outEvent, bool isEof)
+{
+    CHIP_ERROR err                         = CHIP_NO_ERROR;
+    uint16_t maxBlockSize                  = sender.GetTransferBlockSize();
+    System::PacketBufferHandle fakeDataBuf = System::PacketBuffer::NewWithAvailableSize(maxBlockSize);
+    if (fakeDataBuf.IsNull())
+    {
+        NL_TEST_ASSERT(inSuite, false);
+        return;
+    }
+
+    uint8_t * fakeBlockData = fakeDataBuf->Start();
+
+    // Provide Block data and verify sender emits Block message
+    TransferSession::BlockData blockData;
+    blockData.Data   = fakeBlockData;
+    blockData.Length = maxBlockSize;
+    blockData.IsEof  = isEof;
+
+    err = sender.PrepareBlock(blockData);
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+    sender.PollOutput(outEvent);
+    NL_TEST_ASSERT(inSuite, outEvent.EventType == TransferSession::kOutput_MsgToSend);
+    System::PacketBufferHandle blockMsgToSend = outEvent.MsgData.Retain();
+    NL_TEST_ASSERT(inSuite, !blockMsgToSend.IsNull());
+
+    // Pass Block message to receiver and verify Block is received
+    err = receiver.HandleMessageReceived(blockMsgToSend.Retain());
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+    receiver.PollOutput(outEvent);
+    NL_TEST_ASSERT(inSuite, outEvent.EventType == TransferSession::kOutput_BlockReceived);
+    System::PacketBufferHandle receivedBlock = outEvent.MsgData.Retain();
+    NL_TEST_ASSERT(inSuite, !receivedBlock.IsNull());
+    NL_TEST_ASSERT(inSuite, outEvent.blockdata.Data != nullptr);
+}
+
+void SendAndVerifyBlockAck(nlTestSuite * inSuite, void * inContext, TransferSession & sender, TransferSession & receiver,
+                           TransferSession::OutputEvent & outEvent)
+{
+    // Verify PrepareBlockAck() outputs message to send
+    CHIP_ERROR err = receiver.PrepareBlockAck();
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+    receiver.PollOutput(outEvent);
+    NL_TEST_ASSERT(inSuite, outEvent.EventType == TransferSession::kOutput_MsgToSend);
```

Comment:
Suggest adding a way to check that the message, once parsed, is the kind you expect (i.e., that it is actually a Block Ack())

Reply:
will add, thanks